BUG: Improve error message in demux summarize with no data

Oddant1 opened the pull request:
closes #103

Oddant1 commented:
Working on adding tests with empty artifacts

Oddant1 commented:
Changes in https://github.com/qiime2/q2-demux/pull/104/commits/a68e4f2e2686ebd9713e4da1a9c7ccc66afbefc8 were backtracked because upon further inspection of the code they did nothing.

thermokarst commented on q2_demux/_summarize/assets/package-lock.json:
```diff
@@ -0,0 +1,6073 @@
+{
```
Please remove this file from the diff.

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -1171,6 +1172,30 @@ def test_sequence_length_uses_subsample_paired(self):
 
+    def test_empty_single_end(self):
+        empty = SingleLanePerSampleSingleEndFastqDirFmt(
+            self.get_data_path('summarize_empty/empty_SampleDataSingleEnd'),
```
```suggestion
            self.get_data_path('summarize_empty/empty_single_end'),
```

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -1171,6 +1172,30 @@ def test_sequence_length_uses_subsample_paired(self):
+    def test_empty_paired_end_forward(self):
+        empty = SingleLanePerSamplePairedEndFastqDirFmt(
+            self.get_data_path(
+                'summarize_empty/empty_SampleDataPairedEndForward'), mode='r')
```
```suggestion
                'summarize_empty/empty_forward_in_paired_end'), mode='r')
```

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -1171,6 +1172,30 @@ def test_sequence_length_uses_subsample_paired(self):
+    def test_empty_paired_end_reverse(self):
+        empty = SingleLanePerSamplePairedEndFastqDirFmt(
+            self.get_data_path(
+                'summarize_empty/empty_SampleDataPairedEndReverse'), mode='r')
```
```suggestion
                'summarize_empty/empty_reverse_in_paired_end'), mode='r')
```

thermokarst commented on setup.py:
```diff
@@ -36,7 +36,13 @@
+                           'data/filter_samples_paired_end/dir_fmt/*',
+                           'data/summarize_empty/*'
+                           'data/summarize_empty/empty_SampleDataSingleEnd/*',
+                           'data/summarize_empty/'
```
```suggestion
```

thermokarst commented on setup.py:
```diff
@@ -36,7 +36,13 @@
+                           'data/summarize_empty/empty_SampleDataSingleEnd/*',
+                           'data/summarize_empty/'
+                           'empty_SampleDataPairedEndForward/*',
+                           'data/summarize_empty/'
```
```suggestion
```

thermokarst commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -162,6 +162,8 @@ def summarize(output_dir: str, data: _PlotQualView, n: int = 10000) -> None:
         quality_scores, min_seq_len = _subsample_single(sample_map)
 
     forward_scores = pd.DataFrame(quality_scores['forward'])
+    if forward_scores.empty:
```
I think it might make more sense to move this check up higher, say ~L145 ---

```python
if sequence_count == 0:
    raise ValueError('No sequences present.')
```

Or, we could skip the `ValueError` altogether and update the rendering of the viz to skip rendering the quality score plots tab. I think I like that idea more.

Oddant1 replied:
Sorry for my delayed reply, this one kinda slipped through the cracks for me, and I just noticed this latest comment. I don't think the issue referenced in @ChrisKeefe's issue #103 is about having no quality scores, his example artifact contains no reads at all period. Unless I'm misunderstanding something (or put the error message in a place that would imply only the quality scores were missing, which is what it looks like I did) I don't think there would be anything to render at all given his artifact would there? All the same, this may have unearthed a new improvement. If you want I can both improve the error message if there are no reads at all, and simply skip rendering the quality score plots tab if we are only missing quality scores. Would we want to user to be alerted that there were quality scores missing? Or is this `summarize` method also meant to be used on `fasta` files where no quality scores are expected? I'm not really sure how that would work without excepting given the current state of the code though.

Oddant1 replied:
Additionally, the proposed check doesn't catch the case where there are reads in the forward file and no reads in the reverse file. Is this case actually problematic, or did I incorrectly assume it is?

ebolyen replied:
drive-by cc: @thermokarst

thermokarst replied:
@Oddant1 - 

> I don't think the issue referenced in @ChrisKeefe's issue #103 is about having no quality scores, his example artifact contains no reads at all period

Yep, the original issue doesn't reference quality scores, although _your proposed fix does_. What the proposed fix doesn't actually do is check (directly) for an empty table.

> I don't think there would be anything to render at all given his artifact would there?

We can render the first tab of the visualization:

![image](https://user-images.githubusercontent.com/274668/63452380-02a3af00-c3fb-11e9-9e49-4f7277ddf161.png)

Seeing a bunch of "zeros" in there would actually be pretty informative for users.

> If you want I can both improve the error message if there are no reads at all,

I don't think we should raise an Exception --- you can move the warning into the viz itself though (as a banner at the top of the first tab).

> Or is this summarize method also meant to be used on fasta files where no quality scores are expected?

It is not meant for FASTA files (see my note above about quality scores as part of your solution).

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -1171,6 +1172,40 @@ def test_sequence_length_uses_subsample_paired(self):
                 self.assertEqual(payload["minSeqLen"]["reverse"], 5)
 
+    def test_empty_single_end(self):
+        empty = SingleLanePerSampleSingleEndFastqDirFmt(
```
This creates an invalid format (see https://github.com/qiime2/qiime2/issues/490 for more details). Please fix the test data so that it passes the format's `.validate()` method. I suspect this is a problem for all the test data, so please check all test cases' backing data.

thermokarst commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -231,8 +242,9 @@ def summarize(output_dir: str, data: _PlotQualView, n: int = 10000) -> None:
         fh.write(',')
-        forward_stats.to_json(fh)
-        if paired:
+        if not forward_scores.empty:
```
This might break the visualization's [init hook](https://github.com/qiime2/q2-demux/blob/master/q2_demux/_summarize/assets/src/main.js) without a corresponding `else` statement here to ensure that the argument has _some_ value in it.

Oddant1 replied:
Wouldn't it make more sense to change the init hook to `fwdScores = undefined` than to fill it with a placeholder value?

thermokarst replied:
I think "more sense" is a subjective term here --- go with whatever you think works best for you. My concern is this: what happens if badly structured data makes it into this viz? For example, no fwd quality scores, but rev quality scores (I know that example makes no real sense, but, we have seen stranger things). For that example, as you have written it now, there will be too few args for the function, everything will wind up being "offset" in the signature.

Oddant1 replied:
Very true

Oddant1 commented on q2_demux/_summarize/_visualizer.py:
```diff
@@ -231,8 +242,11 @@ def summarize(output_dir: str, data: _PlotQualView, n: int = 10000) -> None:
         fh.write(',')
-        forward_stats.to_json(fh)
-        if paired:
+        if not forward_scores.empty:
```
If `forward_scores` is empty then `forward_stats` doesn't exist

thermokarst replied:
I think this is not very clear, and will confuse future readers of this code. Please refactor to check for the presence of the `forward_stats` variable (that is assuming this check remains, given my request [above](https://github.com/qiime2/q2-demux/pull/104#pullrequestreview-287625579)). You might need to initialize the variable to a `None` value (or something similar).

Oddant1 commented:
@thermokarst don't I already have test data for paired end `empty_fwd` and paired end `empty_rev`? Is there something wrong with it? Additionally, I am unsure of how to check for part of a visualization in a unittest, but I'll figure it out

thermokarst commented:
> @thermokarst don't I already have test data for paired end `empty_fwd` and paired end `empty_rev`?

Yep!

> Is there something wrong with it?

Not to my knowledge. I sent you those files so that you could run the commands yourself.

Manual testing vs unit testing --- I want you to manually test, because I want to make sure you're seeing the big picture w.r.t. the viz.

> Additionally, I am unsure of how to check for part of a visualization in a unittest, but I'll figure it out

Sort out the business logic first, using manual testing, unit testing patterns will become clear once you figure out the business logic.

Oddant1 commented:
I've been doing a lot of testing, and there are some odd things I don't entirely understand going on here. Using your test files, `summarize` ends up with empty `forward_scores` and empty `reverse_scores` regardless of which test file I use. This causes the `interactive_quality_plot` to not render in either case, and I'm not entirely sure how to render it given the fact that there are "allegedly" no reads at all.

thermokarst commented:
> Using your test files, summarize ends up with empty forward_scores and empty reverse_scores regardless of which test file I use.

I noticed that in my testing, too! One thing I haven't looked at yet is how these data files behave in 2019.7, for comparison.

>  and I'm not entirely sure how to render it given the fact that there are "allegedly" no reads at all.

Perhaps look at how these files behave in 2019.7 (or `master`), then see if any of this weird behavior was pre-existing, or, if it was somehow accidentally introduced in this PR. Make sense?

thermokarst commented:
> @thermokarst, these seem to be valid, but it seems like the misspaired situation that we see on the forum once in a while.

Yep, I think we have seen this often enough to put it on the radar.

> Should these actually fail to validate?

Not sure, I have definitely wondered about that, though. I don't know if these cases actually represent invalid data, though. I mean, obviously it would be a silly thing to have this, but trimming tools like cutadapt seem prone to outputting results like this. As well, in cutadapt you have the option of saving the trimmed reads, so I could see some kind of "splitting" action here. Certainly that could all be handled with more refined types though, just rattling off some thoughts.

> Alternatively, maybe we could just prevent fastq from being empty at all?

That would be nice, I think that we have some technical debt related to optional outputs that might be getting in the way. Right now a few actions in the ecosystem rely on being able to output empty fastq (as opposed to being optional).

thermokarst commented:
@Oddant1, following up on an offline discussion I had with @ebolyen here. I am less concerned with the case of missing fwd or rev reads, and more-so interested in us just getting a solid grip on the business logic. I think exploring that business logic through those weird edge cases will be beneficial for getting things locked down. Does that make sense?

Oddant1 commented:
@thermokarst what do you mean by business logic? I'm currently interpreting that to mean the logic behind why we might want to accept files/artifacts like these in the first place, but I'm not sure.

thermokarst commented:
> @thermokarst what do you mean by business logic?

I mean the [flow of data through this action](https://en.wikipedia.org/wiki/Business_logic). How do we structure, guide, and process data from start to finish? Business logic is related to, but different from the _implementation_ of the code. So, the business logic I am concerned with includes things like "why are we looking at the quality scores to determine our next step," and "why do these badly formatted data inputs produce surprising outputs after passing through the viz." It could be the case that it is a mistake in the implementation, but I think it is better to start with clearing up the business rules first, since that governs how the action should actually behave. Make sense?

Oddant1 commented:
What we currently do on master looks something like this

1. Obtain sequences and number of sequences
2. Obtain forward scores
3. Obtain forward stats
4. Repeat the above 2 steps for reverse if we are paired
5. Generate the html
6. Write to the data.jsonp

Notice, that everything aside from this odd little line of code on line 127 assumes forward reads will always exist

___
https://github.com/qiime2/q2-demux/blob/d5dbcb3591d92fb6beff2d2ee7ba77c081907b71/q2_demux/_summarize/_visualizer.py#L123-L127
___

Currently if there are no forward reads we break on step 3 and get that "cannot describe an empty dataframe" deal. On my feature branch, the only change I've really made is to explicitly check for the presence of forward reads in the same way as we explicitly check for the presence of reverse reads. That was the intent anyway.

So it's pretty apparent why we currently explode on lack of forward reads. As for why we don't get any forward scores out of either file, it definitely happens on the master branch not just the feature branch. If we run the command with the .qza with an empty forward, we count 0 sequences, but if we use the one with an empty reverse, we do count a single sequence.

It seems as though we lose this sequence when it comes time to map it to a file, and I think I know why.
 
___
https://github.com/qiime2/q2-demux/blob/d5dbcb3591d92fb6beff2d2ee7ba77c081907b71/q2_demux/_summarize/_visualizer.py#L152-L164
___

On line 153, we call the function `_link_sample_n_to_file`. This next screenshot is what that function does with comments annotating the specific values we will have when the function executes using the `.qza` that has one forward read and no reverse reads.

___
![Selection_099](https://user-images.githubusercontent.com/10642616/64896196-cfd48b80-d633-11e9-9c2d-229faeac2b0e.png)
___

So when we hit the `sample_map` creation on line 157 we get this as our sample map

___
![Selection_100](https://user-images.githubusercontent.com/10642616/64896270-0a3e2880-d634-11e9-924b-9718a4addc76.png)
___

`[forward_file, reverse_file, 0]` telling us we have nothing. I have no idea why `_link_sample_n_to_file` does what it does, but this seems to be the issue with this particular case. It looks like adding more reads MIGHT solve the particular bizarre issue we've run into with it saying there are no reads in this instance, but surely this shouldn't be a thing? It also seems as though this issue has the potential to arise regardless of the number of reads actually present provided the first value in `subsample_ns` is "0". . .

thermokarst commented:
Thanks for this, @Oddant1. I am going to put in ~90 minutes of effort on this to see if I can wrap it up - stay tuned.

thermokarst commented:
Okay @Oddant1, I made a bunch of pretty broad tweaks in 
8f68dc7, let me know if you have any questions. I left a TODO list in that changeset, I hope that will help provide some guidance to run this across this finish line.

Oddant1 commented:
Thanks @thermokarst, sorry I didn't see this earlier. I was tackling a lot of environment related issues for most of today, but they should all be resolved now. I've pulled down your commit, done a first pass through the code, and identified where all the tests are failing and why. I'll get to addressing those todos tomorrow.

Oddant1 commented:
@thermokarst I'm poking around these tests a bit more,  and I just wanted to confirm a few things.

Some of the tests are failing because we no longer separately track minimum sequence length for forward and reverse reads. I already cleaned up the tests, but it occurred to me that I should probably be asking if this behavior is what we want since it seems odd to me to cut something like that (not entirely sure what the significance of it is though)? Should I add the separate checks back to the tests and create a new mechanism by which to track them separately?

Some of the tests are failing because the file `per-sample-fastq-counts.csv` is no longer created. Is this intentional, or do we want that file to still exist?

The code was changed significantly, so it isn't too surprising that some of the tests broke. I just want to make sure the changes to the actual behavior are desired.

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -863,7 +863,7 @@ def test_basic(self):
             self.assertTrue(os.path.exists(index_fp))
             self.assertTrue(os.path.getsize(index_fp) > 0)
-            # NEW TODO: This doesn't get created anymore
+            # NEW TODO: This file is talked about on line 221, but never created?
```
I think it _is_ being created, its just a different extension (TSV in the viz code, but CSV here in the test) - I would start with updating the extensions here in the test suite.

I changed this to TSV (from CSV) because we have been working towards "standardizing" that format across our plugins.

Oddant1 replied:
Yeah I realized a moment ago that I had misread some parts of the changed code last night which prompted that sort of roleback commit. Things are making a lot more sense now.

thermokarst commented on q2_demux/tests/test_demux.py:
```diff
@@ -939,7 +939,7 @@ def test_single_sample_multiple_files(self):
         with tempfile.TemporaryDirectory() as output_dir:
             # TODO: Remove _PlotQualView wrapper
-            # NEW TODO: Not so sure about this one
+            # NEW TODO: Not sure what happened here
```
Just a note, the `TODO` above this, online 941 (`# TODO: Remove _PlotQualView wrapper`) is _super_ old, and does not need to be addressed here.

Oddant1 replied:
Yeah I figure. When I looked through the diff from your commit the only new todos were removing the old functions and the list. I'm labeling all of mine NEW TODO so I can easily distinguish :joy:. Thank you for the clarification though.

thermokarst replied:
We should invent a TODO versioning scheme.

Oddant1 replied:
Automatically detects when new TODOs are added in a commit and changes the versioning appropriately

Oddant1 commented:
Thinking about https://github.com/qiime2/q2-demux/pull/104/commits/68983e4896d67c0d65731c3a0434404b8fe353ec a bit more, I'm not sure this is actually what needs to happen here.I have a different idea coming next commit

Oddant1 commented:
This last test is probably going to be significantly more of a pain to fix. Multiple files are being used with the same sample-id. This is failing when we attempt to transform the manifest to a dataframe because the last step of that transformation wants to pivot the dataframe so the sampleid column is the index, but all three sampleids are "sample1" causing it to break. This may be solvable using `pivot_table` instead of `pivot` in the transfromer. I'll test it and see how it goes.

I also considered the possibility of making the filename an iterable of filenames instead, but this will probably need to be done in all cases (so when there is only one file name) and would probably also need to happen in the transformer. I am open to other suggestions.

The test in question:
https://github.com/Oddant1/q2-demux/blob/74c64f45b2e7fcc90bb74d3a7d3399c040e3b551/q2_demux/tests/test_demux.py#L922-L961

The transformer in question (note, it is in q2-types):
https://github.com/qiime2/q2-types/blob/004403d689f0aa62c54f5041078883773600e304/q2_types/per_sample_sequences/_transformer.py#L390-L400

Reference to `pivot_table` method:
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot_table.html#pandas.DataFrame.pivot_table

thermokarst commented:
> The test in question:

Nuke that test (`test_single_sample_multiple_files`)! It is super specific, and, it shouldn't even be possible to hit this case anymore. I see no need to make your life painful just for this one silly little test (that I wrote three years ago, apparently).

Oddant1 commented:
Works for me :joy:

Oddant1 commented:
As of https://github.com/qiime2/q2-demux/pull/104/commits/5439727f1380333cdce5237da154dbabc037d2eb all tests should pass. I haven't bothered cleaning up some linting errors yet though.

Oddant1 commented:
I think the todo list item 

` - fix "The minimum sequence length identified during subsampling was undefined bases"`

was resolved when I made the `minSeqLen` a dictionary keyed on direction again. The line that reports the `minSeqLen` is created separately for each direction. Once when `plotBoxes` is called with the forward data and again when it's called with the reverse data. This is the line responsible for getting the `minSeqLen` to be put into the note mentioned in the todo.

`  const minSeqLen = seqProps.minSeqLen[data.direction];`

 When minSeqLen was just a single int in https://github.com/qiime2/q2-demux/pull/104/commits/8f68dc7e9cf025f35665f0a773e10a7873a764f4 this line would attempt to index into that integer as if it were a dictionary (or I guess a map or something in Javascript? Idk). This was causing `minSeqLen` to always be undefined. Now the indexing works again.

Oddant1 commented:
I don't really know where to go from here. I can't really think of any new tests that obviously need to be added, and I'm unsure of how to go about consolidating the state variables. There is a lot of state there, but it all seems necessary, and none of it is particularly complicated.

thermokarst commented:
Thanks @Oddant1, I will take a look at your comments and commits above and let you know if there is anything else. Before I do that though, can you fix the linting errors that travis is reporting? Once that's set, please move this to "In Review" and reassign to me. Thanks!

Oddant1 commented:
@thermokarst I cleaned up the non header related linting errors and rebased so the headers are good too. We should pass all tests and have no linting errors on Travis now

thermokarst commented:
Thanks @Oddant1! I made some changes to the JS, so before diving in, don't forget to `make clean && make`). Here are a few laundry-list items:

- [ ] Can you make sure the parametric seven-number summaries haven't changed? Just run the same dataset against this PR, and the 2020.2 release version of this visualizer, and compare the results.
- [ ] Can you make some kind of "placeholder" show up in the quality plot tab when one of the directions is completely missing? Use this test data, please: [test_data.zip](https://github.com/qiime2/q2-demux/files/4635759/test_data.zip)
- [ ] Regarding tests, maybe you can create a few tests using the test data provided above, that checks for:
  - [ ] "placeholders" when a direction are missing (run the plot and look at it for an example)
  - [ ] warnings per direction
  - [ ] other things?

Let me know if you have any questions - thanks!

thermokarst commented on q2_demux/_summarize/assets/package-lock.json:
```diff
@@ -0,0 +1,5686 @@
+{
```
I decided to commit the package-lock.json, but, I might change my mind later.

Oddant1 commented:
Based on the test data in moving pictures the seven number summaries haven't changed

Oddant1 commented:
Looks like we get nothing at all for the artifact with empty forward reads. It says there should be a plot for reverse reads but there's nothing at all.
![Screenshot from 2020-05-15 11-15-07](https://user-images.githubusercontent.com/10642616/82082819-84302100-969d-11ea-97a8-2e43b4ceb8b2.png)

thermokarst commented:
Did you remember to run `make clean && make`, after fetching the latest changes in this branch? This is what the plots look like for me:

### empty-fwd
![Screen Shot 2020-05-15 at 11 18 24 AM](https://user-images.githubusercontent.com/274668/82082992-cfe2ca80-969d-11ea-80f8-999c518c2a76.png)

### empty-rev
![Screen Shot 2020-05-15 at 11 18 59 AM](https://user-images.githubusercontent.com/274668/82083032-e6892180-969d-11ea-9af4-f07480468d98.png)

Oddant1 commented:
I did, but I'll do it again.

Oddant1 commented:
It's working now, maybe I accidentally ran it before pulling the latest commit? Dunno

thermokarst commented:
/shrug - who knows... glad you got it working!

Oddant1 commented:
Ok unfortunately it looks like the seven number summaries did actually change. I must have forgot to pull the commit earlier. This is why I need to not try to do work before drinking coffee :joy:. Looking into potential causes for these discrepancies now while I also look at how to go about adding that placeholder to the viz.

Oddant1 commented:
Also worth noting something I don't think I mentioned earlier. The minSeqLen for the reverse read in one of the tests changed from 5 to 3, I think this has to do with the subsampling looking at different indices than it was before.

thermokarst commented:
> This is why I need to not try to do work before drinking coffee

lol! ☕ <- drink up my friend

> Looking into potential causes for these discrepancies now while I also look at how to go about adding that placeholder to the viz.

Sweet!

> The minSeqLen for the reverse read in one of the tests changed from 5 to 3, I think this has to do with the subsampling looking at different indices than it was before.

This makes sense - I think the old behavior was actually using the global minimum, rather than the direction's minimum length (although I would need to double-check that statement).

Oddant1 commented:
@thermokarst https://github.com/qiime2/q2-demux/pull/104/commits/72d5fe56f6f8602812d717f7c3c133da637b3676 takes a similar approach to the placeholder histograms. It basically just makes the empty div for the quality plot take up the amount of space it would take up if it actually had any information and prints there are no reads in whichever direction is empty. Sorry that took a while. This stuff is fairly unknown to me, so it took a while just to figure out where I needed to be making edits.
